### PR TITLE
Start dune and detect version

### PR DESCRIPTION
**Pull request by rgrinberg**
We attempt to find dune and check that it's recent enough for RPC
functionality. No new functionality is added, this PR is mostly to avoid
conflicts for future PR's.

### CODE REVIEW

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 1]
```diff
@@ -846,18 +847,44 @@ let start () =
...
+              | Out_of_date ->
+                Some
+                  "Dune is out of date. Install dune >= 3.0 for project \
+                   functionality"
+              | _ -> None
+            in
+            match message with
+            | None -> Fiber.return ()
+            (* We disable the warnings because dune 3.0 isn't available yet *)
+            | Some _ when true -> Fiber.return ()
```
Comment (tmattio) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:876]: `Some _ when true`? Did you forget to remove this?

File: ocaml-lsp-server/src/dune.ml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+open Import
+
+type state =
+  | Absent
+  | Out_of_date
+  | Running
```
Comment (tmattio) [inline ocaml-lsp-server/src/dune.ml:6]: What if it's there but is not running, or crashed?

File: ocaml-lsp-server/src/dune.ml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  | Some s -> Fiber.return s
+  | None -> (
+    match Bin.which "dune" with
+    | None ->
+      let+ () = Fiber.Ivar.fill t.state Absent in
+      Absent
+    | Some bin ->
+      let bin = Fpath.to_string bin in
+      let dev_null () = Unix.openfile "/dev/null" [ Unix.O_RDWR ] 0o666 in
+      let stdin = dev_null () in
```
Comment (tmattio) [inline ocaml-lsp-server/src/dune.ml:28]: Did you want to leave out the ability to send requests and receive responses from this PR?

**Comment by rgrinberg** [comment]
Thibaut Mattio ***@***.***> writes:

> `Some _ when true`? Did you forget to remove this?

It's because there's no released version of dune yet, so the warning
would be pesky.

> What if it's there but is not running, or crashed?

Note that "running" refers to the status of the RPC client started by
LSP. So hopefully that should never crash. If dune isn't running, the
rpc client will notify us.

> Did you want to leave out the ability to send requests and receive
> responses from this PR?

Yes. Making actual requests is saved for another PR that will not be
merged in a while. This will require the unreleased client library, so I
don't want master to depend on it just yet.

**Comment by tmattio** [comment]
Thanks for the replies, the intent of the PR is a bit clearer now and it looks good to me 🙂

**Review by tmattio — APPROVED** [review]

File: ocaml-lsp-server/src/dune.mli [hunk 4]
```diff
@@ -0,0 +1,12 @@
+open Import
+
+type state =
+  | Absent
+  | Out_of_date
+  | Running
```
Comment (ulugbekna) [inline ocaml-lsp-server/src/dune.mli:6]: 1. I'm finding name `Absent` slightly confusing, as to it's not very clear where it is absent from, especially in the presence of another variant `Running`. Maybe `Not_found` or `Not_found_on_system` (long but very clear)? 
2. Separate error cases from success cases by using `result`?

File: ocaml-lsp-server/src/dune.mli [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+type t
+
+val state : t -> state Fiber.t
+
+val create : Scheduler.t -> t
```
Comment (ulugbekna) [inline ocaml-lsp-server/src/dune.mli:12]: 3. Isn't it quite unexpected that the function `state` can start a process and mutate things? Yes, we return a fiber, but it still seems suspicious. Wdyt?
4. (minor) While use of `create` would signal that some mutation (launching a process) is happening (vs `make`, for example).

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 6]
```diff
@@ -846,18 +847,44 @@ let start () =
...
+            | Some message ->
+              let* _ = Server.initialized server in
```
Comment (ulugbekna) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:878]: Annotate type for `_`?

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 7]
```diff
@@ -846,18 +847,44 @@ let start () =
...
+            let* state = Dune.state dune in
+            let message =
+              match state with
+              | Absent ->
+                Some "Dune must be installed for project functionality"
+              | Out_of_date ->
+                Some
+                  "Dune is out of date. Install dune >= 3.0 for project \
+                   functionality"
+              | _ -> None
```
Comment (ulugbekna) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:871]: 5. could either expand `_` or separate success/error cases I mentioned in (2)?

File: ocaml-lsp-server/src/dune.mli (hunk 5 above)
Comment (rgrinberg) [reply ocaml-lsp-server/src/dune.mli:12]: 3. It cannot mutate things in a way that is observable to users of the module.
4. That's how I did it initially but it came with its own problems. Like now we need to wait for `create` to complete before we can finish initializing the server.

File: ocaml-lsp-server/src/dune.mli (hunk 4 above)
Comment (rgrinberg) [reply ocaml-lsp-server/src/dune.mli:6]: 1. `Not_found` sounds good to me.
2. It's worth doing if I use any of the functions in `Result`. I also wouldn't call these error states. Just states where the functionality is diminished.

**Pull request merged**